File: Sources/UsageWidget/Domain/ProviderUsage.swift
import Foundation

/// A single snapshot of a provider's usage, returned by `UsageProvider.fetchUsage()`
/// and cached to disk. This is the unified currency the whole app speaks in.
struct ProviderUsage: Hashable, Codable, Sendable {
    let providerID: ProviderID
    var plan: PlanInfo?
    var windows: [LimitWindow]
    var tokens: TokenStats?
    /// When the provider produced this data (used for "updated 3m ago" / staleness).
    var capturedAt: Date

    init(
        providerID: ProviderID,
        plan: PlanInfo? = nil,
        windows: [LimitWindow],
        tokens: TokenStats? = nil,
        capturedAt: Date = Date()
    ) {
        self.providerID = providerID
        self.plan = plan
        self.windows = windows
        self.tokens = tokens
        self.capturedAt = capturedAt
    }

    /// The two windows that drive the hero view, ordered (5-hour first), each present at most once.
    var heroWindows: [LimitWindow] {
        windows.filter { $0.kind.isHero }.sorted { $0.kind.sortIndex < $1.kind.sortIndex }
    }

    /// Non-hero windows (e.g. per-model weekly breakdowns) shown under "Details".
    var detailWindows: [LimitWindow] {
        windows.filter { !$0.kind.isHero }.sorted { $0.kind.sortIndex < $1.kind.sortIndex }
    }

    var fiveHour: LimitWindow? { windows.first { $0.kind == .fiveHour } }
    var weekly: LimitWindow? { windows.first { $0.kind == .weekly } }

    /// Highest utilization across all windows — feeds the menu-bar headline percentage.
    var maxUtilization: Double {
        windows.map(\.clampedUtilization).max() ?? 0
    }
}

File: windows/src/Core/Domain/ProviderUsage.cs
namespace LLMUsageWidget.Core.Domain;

/// <summary>A single snapshot of a provider's usage. The unified currency the whole app speaks in.</summary>
public sealed record ProviderUsage(
    ProviderId ProviderId,
    IReadOnlyList<LimitWindow> Windows,
    PlanInfo? Plan = null,
    TokenStats? Tokens = null,
    DateTimeOffset? CapturedAt = null)
{
    /// <summary>The windows that drive the hero view, ordered (5-hour first).</summary>
    public IReadOnlyList<LimitWindow> HeroWindows =>
        Windows.Where(w => w.Kind.IsHero()).OrderBy(w => w.Kind.SortIndex()).ToList();

    /// <summary>Non-hero windows (e.g. per-model weekly breakdowns) shown under "Details".</summary>
    public IReadOnlyList<LimitWindow> DetailWindows =>
        Windows.Where(w => !w.Kind.IsHero()).OrderBy(w => w.Kind.SortIndex()).ToList();

    public LimitWindow? FiveHour => Windows.FirstOrDefault(w => w.Kind == LimitWindowKind.FiveHour);
    public LimitWindow? Weekly => Windows.FirstOrDefault(w => w.Kind == LimitWindowKind.Weekly);

    /// <summary>Highest utilization across all windows — feeds the menu-bar headline percentage.</summary>
    public double MaxUtilization => Windows.Count == 0 ? 0 : Windows.Max(w => w.ClampedUtilization);
}
